### PR TITLE
Update an Add new Active Directory script to setup app registartion and access

### PR DESCRIPTION
**Pull request by TomMalow**
This PR updates the existing Add-GroupToServicePrincipal.ps1 with the latest breaking changes to Microsoft Graph and az cli. This include a breaking change where API endpoint fails if the assignment already exist. The script now checks if the assignment already exist.

The PR also introduces 3 new scripts:
 - `Add-ClientToServicePrincipal.ps1` - Script to grant clients pre-approved access to an application. Mostly used to grant Visual Studio access to allow for local testing.
 - `Add-MsiAccess.ps1` - Script to add a manage identity access to the current service.
 - `Add-OAuthUserImpersonationScope.ps1` - Script to add user_impersonation scope to app registration. This was previously default behavior when creating an app registration.

### CODE REVIEW

File: azure-cli/ad/Add-OAuthUserImpersonationScope.ps1 [hunk 1]
```diff
@@ -0,0 +1,49 @@
+function Add-OAuthUserImpersonationScope(
+    [Parameter(Mandatory = $true)]
+    [string] $spnId
+) {
+  # import utility functions
+  . "$PSScriptRoot\..\utilities\deploy.naming.ps1"
```
Comment (lupusbytes) [inline azure-cli/ad/Add-OAuthUserImpersonationScope.ps1]: What is used from deploy.naming.ps1 ?

File: azure-cli/ad/Add-OAuthUserImpersonationScope.ps1 [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      adminConsentDisplayName = "Access $($existingManifest.displayName)"
+      id = $([System.guid]::NewGuid().toString())
+      isEnabled = $true
+      type = "User"
+      userConsentDescription = "Allow the application to access $($existingManifest.displayName) on your behalf."
+      userConsentDisplayName = "Access $($existingManifest.displayName)"
+      value = "user_impersonation"
+    }
+
+    $existingPermissionScopes += $newScope
```
Comment (lupusbytes) [inline azure-cli/ad/Add-OAuthUserImpersonationScope.ps1:33]: should this be moved inside the `if ($null -eq $existingScope) { ... }`  scope ?

File: azure-cli/ad/Add-OAuthUserImpersonationScope.ps1 [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    --url https://graph.microsoft.com/v1.0/applications/$spnId `
+    --headers "Content-Type=application/json" `
+    | ConvertFrom-Json
+
+  Throw-WhenError -output $existingManifest
+
+  $existingPermissionScopes = $existingManifest.api.oauth2PermissionScopes
+  $existingScope = $existingPermissionScopes | Where-Object {$_.value -eq "user_impersonation"}
+
+  if ($null -eq $existingScope){
```
Comment (lupusbytes) [inline azure-cli/ad/Add-OAuthUserImpersonationScope.ps1]: nit: missing space after `)`
the whole document can be autoformatted in vscode using alt-shift-f

File: azure-cli/ad/Add-MsiAccess.ps1 [hunk 4]
```diff
@@ -0,0 +1,75 @@
+function Add-MsiAccess(
```
Comment (lupusbytes) [inline azure-cli/ad/Add-MsiAccess.ps1]: Lots of formatting inconsistencies. Use alt-shift-f to autoformat

File: azure-cli/ad/Add-MsiAccess.ps1 [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  [Parameter(Mandatory = $true)]
+  [EnvironmentConfig] $environmentConfig,
```
Comment (lupusbytes) [inline azure-cli/ad/Add-MsiAccess.ps1]: nit: add linebreak after type to keep consistent with rest of script

File: azure-cli/ad/Add-MsiAccess.ps1 [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+  [Parameter(Mandatory = $true)]
+  [NamingConfig] $namingConfig,
```
Comment (lupusbytes) [inline azure-cli/ad/Add-MsiAccess.ps1]: nit: add linebreak after type to keep consistent with rest of script

File: azure-cli/ad/Add-MsiAccess.ps1 [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  $requiredRoles = $null
+) {
+
+  # import utility functions
+  . "$PSScriptRoot\..\utilities\deploy.naming.ps1"
+
+  $appUri = Get-AppIdentityUri `
+    -type $appType `
+    -environmentConfig $environmentConfig `
+    -namingConfig $namingConfig
```
Comment (lupusbytes) [inline azure-cli/ad/Add-MsiAccess.ps1]: add linebreak between the two function invocations to improve readability ?

File: azure-cli/ad/Add-MsiAccess.ps1 [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    --headers 'Content-Type=application/json' `
+    | ConvertFrom-Json).value
+
+  Throw-WhenError -output $existingAssignments
+
+  foreach ($roleValue in $requiredRoles) {
+    foreach ($role in $roles) {
+      if ($roleValue -eq $role.value -or $roleValue -eq "include-all") {
+
+        if ($existingAssignments.appRoleId -contains $($role.id)){
```
Comment (lupusbytes) [inline azure-cli/ad/Add-MsiAccess.ps1]: redundant parentheses on `$($role.id)`. should work with `$role.id`

File: azure-cli/ad/Add-MsiAccess.ps1 [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+          Write-Host "API Permission $($role.value) Already Assigned"
+        } else {
+          Write-Host "Assign API Permission $($role.value)"
+          $output = az rest `
+          --method POST `
+          --uri "https://graph.microsoft.com/v1.0/servicePrincipals/$msiId/appRoleAssignments" `
+          --body "{\`"appRoleId\`": \`"$($role.id)\`",\`"principalId\`": \`"$msiId\`",\`"resourceId\`": \`"$spnId\`"}" `
```
Comment (lupusbytes) [inline azure-cli/ad/Add-MsiAccess.ps1]: are you about the content of the body ?
in my test

```
$role = @{id="1"}
$msiId = "2"
$spnId = "3"
Write-Host "{\`"appRoleId\`": \`"$($role.id)\`",\`"principalId\`": \`"$msiId\`",\`"resourceId\`": \`"$spnId\`"}"
```
produces
`{\"appRoleId\": \"1\",\"principalId\": \"2\",\"resourceId\": \"3\"}`

if you remove the backslashes and keep backticks, you get
`{"appRoleId": "1","principalId": "2","resourceId": "3"}`

File: azure-cli/ad/Add-GroupToServicePrincipal.ps1 [hunk 10]
```diff
@@ -27,10 +27,24 @@ function Add-GroupToServicePrincipal(
...
+
+    if ($existingAssignments.value.principalId -eq $groupId ) {
+      Write-Host "  Group already has access to the Service Principal" -ForegroundColor DarkYellow
+    } else {
+      Write-Host "  Grant group access to Service Principal" -ForegroundColor DarkYellow
+      $output = az rest `
+          --method POST `
+          --url "https://graph.microsoft.com/v1.0/servicePrincipals/$objectId/appRoleAssignedTo" `
+          --headers "Content-Type=application/json" `
+          --body "{\""resourceId\"":\""$objectId\"",\""principalId\"":\""$groupId\""}"
```
Comment (lupusbytes) [inline azure-cli/ad/Add-GroupToServicePrincipal.ps1]: this is the same question as the body in `azure-cli/ad/Add-MsiAccess.ps1` 
are you supposed to send what looks like partially escaped json as a body ?

```
$objectId = "4"
$groupId = "5"
Write-Host "{\""resourceId\"":\""$objectId\"",\""principalId\"":\""$groupId\""}"
```
produces 
`{\"resourceId\":\"4\",\"principalId\":\"5\"}`

File: azure-cli/ad/Add-MsiAccess.ps1 [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  $roles = az ad sp show `
+      --id $spnId `
+      --query appRoles `
+      | ConvertFrom-Json
+
+  Throw-WhenError -output $roles
+
+  if ($null -eq $requiredRoles) {
+    $requiredRoles = @("include-all")
+    Write-Host "Assigning all available roles"
```
Comment (lupusbytes) [inline azure-cli/ad/Add-MsiAccess.ps1]: Should the Write-Host maybe be indented and colored like the other script ?

File: azure-cli/ad/Add-GroupToServicePrincipal.ps1 [hunk 12]
```diff
@@ -27,10 +27,24 @@ function Add-GroupToServicePrincipal(
...
-    --body "{`"resourceId`":`"$objectId`",`"principalId`":`"$GroupId`"}"
+    $existingAssignments = az rest `
+        --method GET `
+        --url "https://graph.microsoft.com/v1.0/servicePrincipals/$objectId/appRoleAssignedTo" `
+        --headers "Content-Type=application/json" ` `
+        | ConvertFrom-Json
+
+    Throw-WhenError -output $existingAssignments
+
+    if ($existingAssignments.value.principalId -eq $groupId ) {
```
Comment (lupusbytes) [inline azure-cli/ad/Add-GroupToServicePrincipal.ps1]: nit: redundant whitespace after `$groupId`

**Review by lupusbytes — CHANGES_REQUESTED** [review]
.

File: azure-cli/ad/Add-OAuthUserImpersonationScope.ps1 (hunk 2 above)
Comment (TomMalow) [reply azure-cli/ad/Add-OAuthUserImpersonationScope.ps1:33]: It already is. The scope ends at line 48

File: azure-cli/ad/Add-OAuthUserImpersonationScope.ps1 (hunk 3 above)
Comment (TomMalow) [reply azure-cli/ad/Add-OAuthUserImpersonationScope.ps1]: Thanks for the tip. Will do that for the other scripts as well

File: azure-cli/ad/Add-OAuthUserImpersonationScope.ps1 (hunk 1 above)
Comment (TomMalow) [reply azure-cli/ad/Add-OAuthUserImpersonationScope.ps1]: Good catch, seems like a copy/paste error.

**Review by lupusbytes — APPROVED** [review]

**Pull request merged**